File: 06_Searching/src/Allocate_Minimum_Pages.js
function findPages(a, n, m) {
    // your code here
    if (m > n) {
        return -1;
    }

    let start = Math.max(...a);
    let end = a.reduce((a,b) => a+b,0);
    let ans = -1;

    while (start <= end) {
        let mid = Math.floor(start + (end-start)/2);

        if (isPossible(a,n,m,mid)) {
            ans = mid;
            end = mid - 1;
        } else {
            start = mid + 1;
        }
    }

    return ans

    function isPossible(a,n,m,mid) {
        let currentPages = a[0];
        let studentRequired = 1;

        for (let i = 1; i < a.length; i++) {
            currentPages += a[i];

            if (currentPages > mid) {
                studentRequired++;
                currentPages = a[i];
            }
        }

        return studentRequired <= m;
    }
}

console.log(findPages([13,31,37,45,46,54,55,63,73,84,85],11,9))

File: 06_Searching/src/Allocate_Minimum_Pages.java
import java.util.Arrays;

public class Allocate_Minimum_Pages {
    public static long findPages(int n, int[] arr, int m) {
        // Your code here
        if (n < m) {
            return -1;
        }

        int start = Arrays.stream(arr).max().getAsInt();
        int end = Arrays.stream(arr).sum();
        long ans = -1;

        while (start <= end) {
            int mid = start + (end-start)/2;

            if (isPossible (n,arr,m,mid)) {
                ans = (long) mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        return ans;
    }

    public static boolean isPossible (int n, int[] arr, int m, int mid) {
        int studentCount = 1;
        int currentPages = arr[0];

        for (int i = 1; i < arr.length; i++) {
            currentPages += arr[i];

            if (currentPages > mid) {
                currentPages = arr[i];
                studentCount++;
            }
        }

        return studentCount <= m;
    }

    public static void main(String[] args) {
        System.out.println(findPages(4,new int[] {12, 34, 67, 90},2) == 113);
        System.out.println(findPages(3,new int[] {15, 17, 20},5) == -1);
        System.out.println(findPages(11,new int[] {13,31,37,45,46,54,55,63,73,84,85},9) == 85);
    }
}
